File: external/java/src/main/java/ch/epfl/dedis/lib/crypto/bn256/BN.java
package ch.epfl.dedis.lib.crypto.bn256;

import java.math.BigInteger;
import java.util.Arrays;
import java.util.Random;

public class BN {

    /**
     * The order of the BN groups G1, G2 and GT.
     */
    public static BigInteger order = Constants.order;

    /**
     * The pair of a G1 point and a scalar that is used to generate it.
     */
    public static class PairG1 {
        private BigInteger k;
        private G1 p;

        public PairG1(BigInteger k, G1 p) {
            this.k = k;
            this.p = p;
        }

        public BigInteger getScalar() {
            return k;
        }

        public G1 getPoint() {
            return p;
        }
    }

    /**
     * The pair of a G2 point and a scalar that is used to generate it.
     */
    public static class PairG2 {
        private BigInteger k;
        private G2 p;

        public PairG2(BigInteger k, G2 p) {
            this.k = k;
            this.p = p;
        }

        public BigInteger getScalar() {
            return k;
        }

        public G2 getPoint() {
            return p;
        }
    }

    /**
     * A point in G1. This object is <em>not</em> thread-safe.
     */
    public static class G1 {
        CurvePoint p;
        public static final int ELEM_SIZE = 256/8;
        public static final int MARSHAL_SIZE = ELEM_SIZE * 2;

        /**
         * Construct a G1 point. There is no guarantee on its value, please set it later.
         */
        public G1() {
            this.p = new CurvePoint();
        }

        /**
         * Construct a G1 point using a given curve point.
         *
         * @param p is the curve point.
         */
        public G1(CurvePoint p) {
            this.p = p;
        }

        /**
         * Copy constructor for G1.
         *
         * @param p point to copy.
         */
        public G1(G1 p) {
            this.p = new CurvePoint(p.p);
        }

        /**
         * Construct a point for a scalar.
         *
         * @param k is the scalar that is multiplied to the generator point to create the object.
         */
        public G1(BigInteger k) {
            this.p = new CurvePoint().mul(CurvePoint.curveGen, k);
        }

        /**
         * Generate a random pair of a point and a scalar that is used to create the point.
         *
         * @param rnd is the random source.
         * @return a pair of a point and a scalar.
         */
        public static PairG1 rand(Random rnd) {
            BigInteger k = randPosBigInt(rnd, Constants.order);
            G1 p = new G1().scalarBaseMul(k);
            return new PairG1(k, p);
        }

        @Override
        public String toString() {
            return "bn256.G1" + this.p.toString();
        }

        /**
         * Perform a scalar multiplication with the generator point.
         *
         * @param k is the scalar.
         * @return the result which is also this object.
         */
        public G1 scalarBaseMul(BigInteger k) {
            this.p.mul(CurvePoint.curveGen, k);
            return this;
        }

        /**
         * Perform a scalar multiplication.
         *
         * @param a is the point.
         * @param k is the scalar.
         * @return the result which is also this object.
         */
        public G1 scalarMul(G1 a, BigInteger k) {
            this.p.mul(a.p, k);
            return this;
        }

        /**
         * Perform a point addition.
         *
         * @param a is a point.
         * @param b is a point.
         * @return the resulting point which is also this object.
         */
        public G1 add(G1 a, G1 b) {
            this.p.add(a.p, b.p);
            return this;
        }

        /**
         * Perform a point negation.
         *
         * @param a is the point for negation.
         * @return the resulting point which is also this object.
         */
        public G1 neg(G1 a) {
            this.p.negative(a.p);
            return this;
        }

        /**
         * Set the point to the infinity (zero) point.
         */
        public G1 setInfinity() {
            this.p.setInfinity();
            return this;
        }

        /**
         * Checks whether the point is an infinity (zero) point.
         *
         * @return true if the point is at infinity.
         */
        public boolean isInfinity() {
            return this.p.isInfinity();
        }

        /**
         * Turns the point into its byte representation.
         *
         * @return the marshalled bytes.
         */
        public byte[] marshal() {
            // operate on a copy so that we do not modify the underlying curve during marshal
            BN.G1 c = new BN.G1(this);

            if (c.p.isInfinity()) {
                return new byte[MARSHAL_SIZE];
            }

            c.p.makeAffine();

            byte[] xBytes = bigIntegerToBytes(c.p.x.mod(Constants.p));
            byte[] yBytes = bigIntegerToBytes(c.p.y.mod(Constants.p));

            byte[] ret = new byte[MARSHAL_SIZE];
            System.arraycopy(xBytes, 0, ret, 1 * ELEM_SIZE - xBytes.length, xBytes.length);
            System.arraycopy(yBytes, 0, ret, 2 * ELEM_SIZE - yBytes.length, yBytes.length);

            return ret;
        }

        /**
         * Turns the byte representation to the point.
         *
         * @param m the input bytes.
         * @return an unmarshalled point when successful, otherwise null.
         */
        public G1 unmarshal(byte[] m) {
            if (m.length != MARSHAL_SIZE) {
                return null;
            }

            this.p.x = new BigInteger(1, Arrays.copyOfRange(m, 0 * ELEM_SIZE, 1 * ELEM_SIZE));
            this.p.y = new BigInteger(1, Arrays.copyOfRange(m, 1 * ELEM_SIZE, 2 * ELEM_SIZE));

            if (this.p.x.signum() == 0 && this.p.y.signum() == 0) {
                this.p.y = BigInteger.ONE;
                this.p.z = BigInteger.ZERO;
                this.p.t = BigInteger.ZERO;
            } else {
                this.p.z = BigInteger.ONE;
                this.p.t = BigInteger.ONE;
                if (!this.p.isOnCurve()) {
                    return null;
                }
            }

            return this;
        }
    }

    /**
     * A point in G2. This object is <em>not</em> thread-safe.
     */
    public static class G2 {
        TwistPoint p;
        public static final int ELEM_SIZE = 256 / 8;
        public static final int MARSHAL_SIZE = ELEM_SIZE * 4;


        /**
         * Construct a G2 point. We make no guarantee on its value, please set it later.
         */
        public G2() {
            this.p = new TwistPoint();
        }

        /**
         * Construct a G2 point from an existing TwistPoint.
         *
         * @param p is the twist point.
         */
        public G2(TwistPoint p) {
            this.p = p;
        }

        /**
         * Copy construct for G2.
         *
         * @param p is G2 point to be copied.
         */
        public G2(G2 p) {
            this.p = new TwistPoint(p.p);
        }

        /**
         * Construct a point for a scalar.
         *
         * @param k is the scalar that is multiplied to the generator point to create the object.
         */
        public G2(BigInteger k) {
            this.p = new TwistPoint().mul(TwistPoint.twistGen, k);
        }

        /**
         * Generate a random pair of a point and a scalar that is used to create the point.
         *
         * @param rnd is the random source.
         * @return a pair of a point and a scalar.
         */
        public static PairG2 rand(Random rnd) {
            BigInteger k = randPosBigInt(rnd, Constants.order);
            G2 p = new G2().scalarBaseMul(k);
            return new PairG2(k, p);
        }

        @Override
        public String toString() {
            return "bn256.G2" + this.p.toString();
        }

        /**
         * Perform a scalar multiplication with the generator point.
         *
         * @param k is the scalar.
         * @return the result which is also this object.
         */
        public G2 scalarBaseMul(BigInteger k) {
            this.p.mul(TwistPoint.twistGen, k);
            return this;
        }

        /**
         * Perform a scalar multiplication.
         *
         * @param a is the point.
         * @param k is the scalar.
         * @return the result which is also this object.
         */
        public G2 scalarMul(G2 a, BigInteger k) {
            this.p.mul(a.p, k);
            return this;
        }

        /**
         * Perform a point addition.
         *
         * @param a is a point.
         * @param b is a point.
         * @return the resulting point which is also this object.
         */
        public G2 add(G2 a, G2 b) {
            this.p.add(a.p, b.p);
            return this;
        }

        /**
         * Perform a point negation.
         *
         * @param a is the point to negate.
         * @return the resulting point which is also this object.
         */
        public G2 neg(G2 a) {
            this.p.negative(a.p);
            return this;
        }

        /**
         * Set the point to the infinity (zero) point.
         */
        public G2 setInfinity() {
            this.p.setInfinity();
            return this;
        }

        /**
         * Checks whether the point is an infinity (zero) point.
         *
         * @return true if the point is at infinity.
         */
        public boolean isInfinity() {
            return this.p.isInfinity();
        }

        /**
         * Turns the point into its byte representation.
         *
         * @return the marshalled bytes.
         */
        public byte[] marshal() {
            if (this.p.isInfinity()) {
                return new byte[MARSHAL_SIZE];
            }

            // operate on a copy so that we do not modify the underlying curve during marshal
            BN.G2 c = new BN.G2(this);

            c.p.makeAffine();

            byte[] xxBytes = bigIntegerToBytes(c.p.x.x.mod(Constants.p));
            byte[] xyBytes = bigIntegerToBytes(c.p.x.y.mod(Constants.p));
            byte[] yxBytes = bigIntegerToBytes(c.p.y.x.mod(Constants.p));
            byte[] yyBytes = bigIntegerToBytes(c.p.y.y.mod(Constants.p));

            byte[] ret = new byte[MARSHAL_SIZE];
            System.arraycopy(xxBytes, 0, ret, 1 * ELEM_SIZE - xxBytes.length, xxBytes.length);
            System.arraycopy(xyBytes, 0, ret, 2 * ELEM_SIZE - xyBytes.length, xyBytes.length);
            System.arraycopy(yxBytes, 0, ret, 3 * ELEM_SIZE - yxBytes.length, yxBytes.length);
            System.arraycopy(yyBytes, 0, ret, 4 * ELEM_SIZE - yyBytes.length, yyBytes.length);

            return ret;
        }

        /**
         * Turns the byte representation to the point.
         *
         * @param m the input bytes.
         * @return an unmarshalled point when successful, otherwise null.
         */
        public G2 unmarshal(byte[] m) {
            if (m.length != MARSHAL_SIZE) {
                return null;
            }

            if (this.p == null) {
                this.p = new TwistPoint();
            }

            this.p.x.x = new BigInteger(1, Arrays.copyOfRange(m, 0 * ELEM_SIZE, 1 * ELEM_SIZE));
            this.p.x.y = new BigInteger(1, Arrays.copyOfRange(m, 1 * ELEM_SIZE, 2 * ELEM_SIZE));
            this.p.y.x = new BigInteger(1, Arrays.copyOfRange(m, 2 * ELEM_SIZE, 3 * ELEM_SIZE));
            this.p.y.y = new BigInteger(1, Arrays.copyOfRange(m, 3 * ELEM_SIZE, 4 * ELEM_SIZE));

            if (this.p.x.x.signum() == 0 && this.p.x.y.signum() == 0 && this.p.y.x.signum() == 0 && this.p.y.y.signum() == 0) {
                this.p.y.setOne();
                this.p.z.setZero();
                this.p.t.setZero();
            } else {
                this.p.z.setOne();
                this.p.t.setOne();

                if (!this.p.isOnCurve()) {
                    return null;
                }
            }

            return this;
        }
    }

    /**
     * GT represents an element in the GT field. This object is <em>not</em> thread-safe.
     */
    public static class GT {
        GFp12 p;
        public static final int ELEM_SIZE = 256 / 8;
        public static final int MARSHAL_SIZE = ELEM_SIZE * 12;

        /**
         * Construct a new GT object, we make no guarantee on its value, please set it later.
         */
        public GT() {
            this.p = new GFp12();
        }

        /**
         * Construct a new GT object from a GFp12 object.
         *
         * @param p is the GFp12 object.
         */
        public GT(GFp12 p) {
            this.p = p;
        }

        @Override
        public String toString() {
            return "bn256.GT" + this.p.toString();
        }

        @Override
        public boolean equals(Object obj)  {
            if (obj == null) {
                return false;
            }
            if (!(obj instanceof BN.GT)) {
                return false;
            }
            BN.GT other = (BN.GT)obj;
            return other.p.equals(this.p);
        }

        /**
         * Perform a scalar multiplication.
         *
         * @param a is the point.
         * @param k is the scalar.
         * @return the result which is also this object.
         */
        public GT scalarMul(GT a, BigInteger k) {
            this.p.exp(a.p, k);
            return this;
        }

        /**
         * Perform an addition.
         *
         * @param a is an element.
         * @param b is an element.
         * @return the resulting element which is also this object.
         */
        public GT add(GT a, GT b) {
            this.p.mul(a.p, b.p);
            return this;
        }

        /**
         * Perform a negation.
         *
         * @param a is the element for negation.
         * @return the resulting element which is also this object.
         */
        public GT neg(GT a) {
            this.p.invert(a.p);
            return this;
        }

        /**
         * Turns the element into its byte representation.
         *
         * @return the marshalled bytes.
         */
        public byte[] marshal() {
            this.p.minimal();

            byte[] xxxBytes = bigIntegerToBytes(this.p.x.x.x);
            byte[] xxyBytes = bigIntegerToBytes(this.p.x.x.y);
            byte[] xyxBytes = bigIntegerToBytes(this.p.x.y.x);
            byte[] xyyBytes = bigIntegerToBytes(this.p.x.y.y);
            byte[] xzxBytes = bigIntegerToBytes(this.p.x.z.x);
            byte[] xzyBytes = bigIntegerToBytes(this.p.x.z.y);
            byte[] yxxBytes = bigIntegerToBytes(this.p.y.x.x);
            byte[] yxyBytes = bigIntegerToBytes(this.p.y.x.y);
            byte[] yyxBytes = bigIntegerToBytes(this.p.y.y.x);
            byte[] yyyBytes = bigIntegerToBytes(this.p.y.y.y);
            byte[] yzxBytes = bigIntegerToBytes(this.p.y.z.x);
            byte[] yzyBytes = bigIntegerToBytes(this.p.y.z.y);

            byte[] ret = new byte[MARSHAL_SIZE];
            System.arraycopy(xxxBytes, 0, ret, 1 * ELEM_SIZE - xxxBytes.length, xxxBytes.length);
            System.arraycopy(xxyBytes, 0, ret, 2 * ELEM_SIZE - xxyBytes.length, xxyBytes.length);
            System.arraycopy(xyxBytes, 0, ret, 3 * ELEM_SIZE - xyxBytes.length, xyxBytes.length);
            System.arraycopy(xyyBytes, 0, ret, 4 * ELEM_SIZE - xyyBytes.length, xyyBytes.length);
            System.arraycopy(xzxBytes, 0, ret, 5 * ELEM_SIZE - xzxBytes.length, xzxBytes.length);
            System.arraycopy(xzyBytes, 0, ret, 6 * ELEM_SIZE - xzyBytes.length, xzyBytes.length);
            System.arraycopy(yxxBytes, 0, ret, 7 * ELEM_SIZE - yxxBytes.length, yxxBytes.length);
            System.arraycopy(yxyBytes, 0, ret, 8 * ELEM_SIZE - yxyBytes.length, yxyBytes.length);
            System.arraycopy(yyxBytes, 0, ret, 9 * ELEM_SIZE - yyxBytes.length, yyxBytes.length);
            System.arraycopy(yyyBytes, 0, ret, 10 * ELEM_SIZE - yyyBytes.length, yyyBytes.length);
            System.arraycopy(yzxBytes, 0, ret, 11 * ELEM_SIZE - yzxBytes.length, yzxBytes.length);
            System.arraycopy(yzyBytes, 0, ret, 12 * ELEM_SIZE - yzyBytes.length, yzyBytes.length);

            return ret;
        }

        /**
         * Turns the byte representation to the element.
         *
         * @param m the input bytes.
         * @return an unmarshalled element when successful, otherwise null.
         */
        public GT unmarshal(byte[] m) {
            if (m.length != MARSHAL_SIZE) {
                return null;
            }

            if (this.p == null) {
                this.p = new GFp12();
            }

            this.p.x.x.x = new BigInteger(1, Arrays.copyOfRange(m, 0 * ELEM_SIZE, 1 * ELEM_SIZE));
            this.p.x.x.y = new BigInteger(1, Arrays.copyOfRange(m, 1 * ELEM_SIZE, 2 * ELEM_SIZE));
            this.p.x.y.x = new BigInteger(1, Arrays.copyOfRange(m, 2 * ELEM_SIZE, 3 * ELEM_SIZE));
            this.p.x.y.y = new BigInteger(1, Arrays.copyOfRange(m, 3 * ELEM_SIZE, 4 * ELEM_SIZE));
            this.p.x.z.x = new BigInteger(1, Arrays.copyOfRange(m, 4 * ELEM_SIZE, 5 * ELEM_SIZE));
            this.p.x.z.y = new BigInteger(1, Arrays.copyOfRange(m, 5 * ELEM_SIZE, 6 * ELEM_SIZE));
            this.p.y.x.x = new BigInteger(1, Arrays.copyOfRange(m, 6 * ELEM_SIZE, 7 * ELEM_SIZE));
            this.p.y.x.y = new BigInteger(1, Arrays.copyOfRange(m, 7 * ELEM_SIZE, 8 * ELEM_SIZE));
            this.p.y.y.x = new BigInteger(1, Arrays.copyOfRange(m, 8 * ELEM_SIZE, 9 * ELEM_SIZE));
            this.p.y.y.y = new BigInteger(1, Arrays.copyOfRange(m, 9 * ELEM_SIZE, 10 * ELEM_SIZE));
            this.p.y.z.x = new BigInteger(1, Arrays.copyOfRange(m, 10 * ELEM_SIZE, 11 * ELEM_SIZE));
            this.p.y.z.y = new BigInteger(1, Arrays.copyOfRange(m, 11 * ELEM_SIZE, 12 * ELEM_SIZE));

            return this;
        }
    }

    /**
     * Perform the pairing operation.
     *
     * @param g1 is the G1 point.
     * @param g2 is the G2 point.
     * @return the GT point.
     */
    public static GT pair(G1 g1, G2 g2) {
        return new GT(OptAte.optimalAte(g2.p, g1.p));
    }

    static BigInteger randPosBigInt(Random rnd, BigInteger n) {
        BigInteger r;
        do {
            r = new BigInteger(n.bitLength(), rnd);
        } while (r.signum() <= 0 || r.compareTo(n) >= 0);
        return r;
    }

    /**
     * We have to use this function instead of the BigInteger.toByteArray method because the latter might produce
     * a leading zero which is different from the Go implementation.
     */
    static byte[] bigIntegerToBytes(final BigInteger a) {
        byte[] bytes = a.toByteArray();
        if (bytes[0] == 0) {
            return Arrays.copyOfRange(bytes, 1, bytes.length);
        }
        return bytes;
    }
}

File: external/js/kyber/src/pairing/bn.ts
import BN from 'bn.js';
import CurvePoint from './curve-point';
import TwistPoint from './twist-point';
import GfP2 from './gfp2';
import GfP12 from './gfp12';
import GfP6 from './gfp6';
import { optimalAte } from './opt-ate';

export type BNType = number | string | number[] | Buffer | BN;

/**
 * Wrapper around the basic curve point. It acts as a mutable object and
 * then every modification is done in-place.
 */
export class G1 {
    /**
     * Hash the message to a point
     * @param msg The message to hash
     * @returns a valid point
     */
    public static hashToPoint(msg: Buffer): G1 {
        const g1 = new G1();
        g1.p = CurvePoint.hashToPoint(msg);

        return g1;
    }

    private static ELEM_SIZE = 256/8;
    private static MARSHAL_SIZE = G1.ELEM_SIZE * 2;

    private p: CurvePoint;

    constructor(k?: BNType) {
        this.p = new CurvePoint();

        if (k) {
            this.scalarBaseMul(new BN(k));
        }
    }

    /**
     * Get the curve point
     * @returns the point
     */
    getPoint(): CurvePoint {
        return this.p;
    }

    /**
     * Set the point to the generator of the curve
     */
    setBase(): void {
        this.p = CurvePoint.generator.clone();
    }

    /**
     * Set the point to infinity
     */
    setInfinity(): void {
        this.p.setInfinity();
    }

    /**
     * Check if the point is the infinity
     * @returns true when infinity, false otherwise
     */
    isInfinity(): boolean {
        return this.p.isInfinity();
    }

    /**
     * Multiply the generator by the scalar k and set the value
     * @param k the scalar
     */
    scalarBaseMul(k: BN): void {
        this.p.mul(CurvePoint.generator, k);
    }

    /**
     * Multiply a by the scalar k and set the value
     * @param a the point
     * @param k the scalar
     */
    scalarMul(a: G1, k: BN): void {
        this.p.mul(a.p, k);
    }

    /**
     * Add a to b and set the value
     * @param a the first point
     * @param b the second point
     */
    add(a: G1, b: G1): void {
        this.p.add(a.p, b.p);
    }

    /**
     * Compute the negative of a and set the value
     * @param the point to negate
     */
    neg(a: G1): void {
        this.p.negative(a.p);
    }

    /**
     * Get the buffer size after marshaling
     * @returns the length
     */
    marshalSize(): number {
        return G1.MARSHAL_SIZE;
    }

    /**
     * Serialize the point into bytes
     * @returns the buffer
     */
    marshal(): Buffer {
        const p = this.p.clone();
        const buf = Buffer.alloc(G1.MARSHAL_SIZE, 0);

        if (p.isInfinity()) {
            return buf;
        }

        p.makeAffine();

        const xBytes = p.getX().toBytes();
        const yBytes = p.getY().toBytes();

        return Buffer.concat([xBytes, yBytes]);
    }

    /**
     * Take a buffer to deserialize a point
     * @param bytes the buffer
     */
    unmarshal(bytes: Buffer): void {
        if (bytes.length != G1.MARSHAL_SIZE) {
            throw new Error('wrong buffer size for a G1 point');
        }

        this.p = new CurvePoint(bytes.slice(0, G1.ELEM_SIZE), bytes.slice(G1.ELEM_SIZE), 1, 1);

        if (this.p.getX().isZero() && this.p.getY().isZero()) {
            this.p.setInfinity();
            return;
        }

        if (!this.p.isOnCurve()) {
            throw new Error('malformed G1 point');
        }
    }

    /**
     * Check the equality between the point and the object
     * @param o the object
     * @returns true when both are equal, false otherwise
     */
    equals(o: any): o is G1 {
        if (!(o instanceof G1)) {
            return false;
        }

        return this.p.equals(o.p);
    }

    /**
     * Get a clone of the element
     * @returns the new element
     */
    clone(): G1 {
        const g = new G1();
        g.p = this.p.clone();
        return g;
    }

    /**
     * Get the string representation of the point
     * @returns the string representation
     */
    toString(): string {
        return `bn256.G1${this.p.toString()}`;
    }
}

/**
 * Wrapper around the twist point. It acts as a mutable object and
 * then every modification is done in-place.
 */
export class G2 {
    private static ELEM_SIZE = 256 / 8;
    private static MARSHAL_SIZE = G2.ELEM_SIZE * 4;

    private p: TwistPoint;

    constructor(k?: BNType) {
        this.p = new TwistPoint();

        if (k) {
            this.scalarBaseMul(new BN(k));
        }
    }

    /**
     * Get the twist point
     * @returns the point
     */
    getPoint(): TwistPoint {
        return this.p;
    }

    /**
     * Set to the generator of the curve
     */
    setBase(): void {
        this.p = TwistPoint.generator.clone();
    }

    /**
     * Set the point to the infinity
     */
    setInfinity(): void {
        this.p.setInfinity();
    }

    /**
     * Check if the point is the infinity
     * @returns true when infinity, false otherwise
     */
    isInfinity(): boolean {
        return this.p.isInfinity();
    }

    /**
     * Mutliply the generator by a scalar k and set the value
     * @param k the scalar
     */
    scalarBaseMul(k?: BN): void {
        this.p.mul(TwistPoint.generator, k);
    }

    /**
     * Multiply a by a scalar k and set the value
     * @param a the point
     * @param k the scalar
     */
    scalarMul(a: G2, k: BN): void {
        this.p.mul(a.p, k);
    }

    /**
     * Add a to b and set the value
     * @param a the first point
     * @param b the second point
     */
    add(a: G2, b: G2): void {
        this.p.add(a.p, b.p);
    }

    /**
     * Compute the negative of a and set the value
     * @param a the point
     */
    neg(a: G2) {
        this.p.neg(a.p);
    }

    /**
     * Get the size of the buffer after marshaling
     * @returns the size
     */
    marshalSize(): number {
        return G2.MARSHAL_SIZE;
    }

    /**
     * Serialize the point into bytes
     * @returns the buffer
     */
    marshal(): Buffer {
        if (this.isInfinity()) {
            return Buffer.alloc(G2.MARSHAL_SIZE, 0);
        }

        const t = this.clone();
        t.p.makeAffine();

        const xxBytes = t.p.getX().getX().toBytes();
        const xyBytes = t.p.getX().getY().toBytes();
        const yxBytes = t.p.getY().getX().toBytes();
        const yyBytes = t.p.getY().getY().toBytes();

        return Buffer.concat([xxBytes, xyBytes, yxBytes, yyBytes]);
    }

    /**
     * Take a buffer and deserialize a point
     * @param bytes the buffer
     */
    unmarshal(bytes: Buffer): void {
        if (bytes.length !== G2.MARSHAL_SIZE) {
            throw new Error('wrong buffer size for G2 point');
        }

        const xxBytes = bytes.slice(0, G2.ELEM_SIZE);
        const xyBytes = bytes.slice(G2.ELEM_SIZE, G2.ELEM_SIZE * 2);
        const yxBytes = bytes.slice(G2.ELEM_SIZE * 2, G2.ELEM_SIZE * 3);
        const yyBytes = bytes.slice(G2.ELEM_SIZE * 3);

        this.p = new TwistPoint(
            new GfP2(xxBytes, xyBytes),
            new GfP2(yxBytes, yyBytes),
            GfP2.one(),
            GfP2.one(),
        );

        if (this.p.getX().isZero() && this.p.getY().isZero()) {
            this.p.setInfinity();
            return;
        }
        
        if (!this.p.isOnCurve()) {
            throw new Error('malformed G2 point');
        }
    }

    /**
     * Get a clone of G2
     * @returns the clone
     */
    clone(): G2 {
        const t = new G2();
        t.p = this.p.clone();

        return t;
    }

    /**
     * Check the equality of the current point and the object
     * @param o the object
     * @returns true when both are equal, false otherwise
     */
    equals(o: any): o is G2 {
        if (!(o instanceof G2)) {
            return false;
        }

        return this.p.equals(o.p);
    }

    /**
     * Get the string representation of the point
     * @returns the string representation
     */
    toString(): string {
        return `bn256.G2${this.p.toString()}`;
    }
}

/**
 * Wrapper around the result of pairing of G1 and G2. It acts as a mutable
 * object and then every modification is done in-place.
 */
export class GT {
    private static ELEM_SIZE = 256 / 8;
    private static MARSHAL_SIZE = GT.ELEM_SIZE * 12;

    public static pair(g1: G1, g2: G2): GT {
        return optimalAte(g1, g2);
    }

    public static one(): GT {
        return new GT(GfP12.one());
    }

    private g: GfP12;

    constructor(g?: GfP12) {
        this.g = g || new GfP12();
    }

    /**
     * Check if the point is one
     * @returns true when one, false otherwise
     */
    isOne(): boolean {
        return this.g.isOne();
    }

    /**
     * Multiply the point a by a scalar k and set the value
     * @param a the point
     * @param k the scalar
     */
    scalarMul(a: GT, k: BN): void {
        this.g = a.g.exp(k);
    }

    /**
     * Add two points a and b and set the value
     * @param a the first point
     * @param b the second point
     */
    add(a: GT, b: GT): void {
        this.g = a.g.mul(b.g);
    }

    /**
     * Compute the negative of a and set the value
     * @param a the point
     */
    neg(a: GT): void {
        this.g = a.g.invert();
    }

    /**
     * Serialize the point into bytes
     * @returns the buffer
     */
    marshal(): Buffer {
        const xxxBytes = this.g.getX().getX().getX().toBytes();
        const xxyBytes = this.g.getX().getX().getY().toBytes();
        const xyxBytes = this.g.getX().getY().getX().toBytes();
        const xyyBytes = this.g.getX().getY().getY().toBytes();
        const xzxBytes = this.g.getX().getZ().getX().toBytes();
        const xzyBytes = this.g.getX().getZ().getY().toBytes();
        const yxxBytes = this.g.getY().getX().getX().toBytes();
        const yxyBytes = this.g.getY().getX().getY().toBytes();
        const yyxBytes = this.g.getY().getY().getX().toBytes();
        const yyyBytes = this.g.getY().getY().getY().toBytes();
        const yzxBytes = this.g.getY().getZ().getX().toBytes();
        const yzyBytes = this.g.getY().getZ().getY().toBytes();

        return Buffer.concat([
            xxxBytes, xxyBytes, xyxBytes,
            xyyBytes, xzxBytes, xzyBytes,
            yxxBytes, yxyBytes, yyxBytes,
            yyyBytes, yzxBytes, yzyBytes,
        ]);
    }

    /**
     * Take a buffer and deserialize a point
     * @param bytes the buffer
     */
    unmarshal(bytes: Buffer): void {
        if (bytes.length !== GT.MARSHAL_SIZE) {
            throw new Error('wrong buffer size for a GT point');
        }

        const n = GT.ELEM_SIZE;
        const xxxBytes = bytes.slice(0, n);
        const xxyBytes = bytes.slice(n, n*2);
        const xyxBytes = bytes.slice(n*2, n*3);
        const xyyBytes = bytes.slice(n*3, n*4);
        const xzxBytes = bytes.slice(n*4, n*5);
        const xzyBytes = bytes.slice(n*5, n*6);
        const yxxBytes = bytes.slice(n*6, n*7);
        const yxyBytes = bytes.slice(n*7, n*8);
        const yyxBytes = bytes.slice(n*8, n*9);
        const yyyBytes = bytes.slice(n*9, n*10);
        const yzxBytes = bytes.slice(n*10, n*11);
        const yzyBytes = bytes.slice(n*11);

        this.g = new GfP12(
            new GfP6(new GfP2(xxxBytes, xxyBytes), new GfP2(xyxBytes, xyyBytes), new GfP2(xzxBytes, xzyBytes)),
            new GfP6(new GfP2(yxxBytes, yxyBytes), new GfP2(yyxBytes, yyyBytes), new GfP2(yzxBytes, yzyBytes)),
        );
    }

    /**
     * Check the equality of the point and an object
     * @param o the object
     * @returns true when both are equal, false otherwise
     */
    equals(o: any): o is GT {
        if (!(o instanceof GT)) {
            return false;
        }

        return this.g.equals(o.g);
    }

    /**
     * Get the string representation of the point
     * @returns the string representation
     */
    toString(): string {
        return `bn256.GT${this.g.toString()}`;
    }
}
